build_tools/lint.fish

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env fish
+#
+# This is meant to be run by "make lint" or "make lint-all". It is not meant to
+# be run directly from a shell prompt.
+#
+set cppchecks warning,performance,portability,information,missingInclude
+set cppcheck_args
+set c_files
+set all no
+
+set -gx CXX $argv[1]
+set -e argv[1]
+
+if test "$argv[1]" = "--all"
+    set all yes
+    set c_files src/
+    set cppchecks "$cppchecks,unusedFunction"
+    set -e argv[1]
+end
+
+# We only want -D and -I options to be passed thru to cppcheck.
+for arg in $argv
+    if string match -q -- '-D*' $arg
+        set cppcheck_args $cppcheck_args $arg
+    else if string match -q -- '-I*' $arg
+        set cppcheck_args $cppcheck_args $arg
+    end
+end
+if test (uname -m) = "x86_64"
+    set cppcheck_args -D__x86_64__ -D__LP64__ $cppcheck_args
+end
+
+if test $all = no
+    # We haven't been asked to lint all the source. If there are uncommitted
+    # changes lint those, else lint the files in the most recent commit.
+    set pending (git status --short --untracked-files=all | sed -e 's/^ *//')
+    if count $pending > /dev/null
+        # There are pending changes so lint those files.
+        for arg in $pending
+            set files $files (string split -m 1 ' ' $arg)[2]
+        end
+    else
+        # No pending changes so lint the files in the most recent commit.
+        set files (git show --name-only --pretty=oneline head | tail --lines=+2)
+    end
+
+    # Filter out the non-C/C++ files.
+    for file in $files
+        if string match -q -- '*.cpp' $file
+            set c_files $c_files $file
+        else if string match -q -- '*.c' $file
+            set c_files $c_files $file
+        end
+    end
+else
+    set c_files src/*.cpp
+end
+
+# We now have a list of files to check so run the linters.
+if set -q c_files[1]
+    if command -s cppcheck > /dev/null
```

Comment:
Hahaha! I made the same suggestion a few weeks ago on someone else's change (zanchey's?) and they pointed out the `type` is a function that calls `command` so better to just use the latter. Personally I think the `type -q` formulation is clearer and the extra overhead is probably not even measurable in most scripts. Done.